consumer/consumererror/README.md

```diff
@@ -0,0 +1,115 @@
+# Consumer errors
+
+This package contains error types that should be returned by a consumer when an
+error occurs while processing telemetry. The error types included in this
+package provide functionality for communicating details about
+the error for use upstream in the pipeline. Ideally the error returned by a
+component in its `consume` function should be from this package.
+
+## Use cases
+
+**Retry logic**: Errors should be allowed to embed rejected telemetry items to
+be retried by the consumer.
+
+**Indicating partial success**: Errors can indicate that not all items were
+accepted, for example as in an OTLP partial success message.
+
+**Communicating network error codes**: Errors should allow embedding information
+necessary for the Collector to act as a proxy for a backend, i.e. relay a status
+code returned from a backend in a response to a system upstream from the
+Collector.
+
+## Creating Errors
+
+Errors can be created through one of the three constructors provided:
+
+- `consumererror.New[Signal]` for retryable errors.
+- `consumererror.NewPartial` for errors where only some of the items failed.
+- `consumererror.NewHTTPStatus` for errors resulting from an HTTP call with an error status code.
+- `consumererror.NewGRPCStatus` for errors resulting from a gRPC call with an error status code.
+
+Only retryable errors will be retried, all other errors are considered permanent
+and will not be retried.
+
+Errors can be joined by passing them to a call to `errors.Join`.
+
+
+### Per-signal retryable errors
+
+If an error is considered transient and data processing should be retried, the data
+that should be retried should be created using `NewTraces`, `NewMetrics` or `NewLogs`
+with the data that should be retried, e.g. `NewTraces(err, traces)`. The data that
+is contained in the error will be the only data that is retried, the caller should
+not use its copy of the data for the retry.
+
+## Other considerations
+
+To keep error analysis simple when looking at an error upstream in a pipeline,
+the component closest to the source of an error or set of errors should make a
+decision about the nature of the error. The following are a few places where
+special considerations may need to be made.
+
+### Fanouts
+
+#### Permanent errors
+
+When a fanout component receives multiple permanent errors from downstream in
+the pipeline, it should determine the error with the highest precedence and
+forward that upward in the pipeline. If a precedence cannot be determined,
+the component should forward all errors.
+
+#### Retryable errors
+
+When a fanout component receives a retryable error, it must only retry the data
+the failed pipeline branch.
```

Comment:
Fanout doesn't retry itself. So it must include some information about the error destination if we want to achieve this

Reply:
The fanout consumer doesn't retry itself, but I think it has the necessary context to make the retries since it knows which component it sends to in its [consume function](https://github.com/open-telemetry/opentelemetry-collector/blob/main/internal/fanoutconsumer/traces.go#L47). Otherwise we can do retries on individual pipeline branches like in the [exporter helper](https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/exporterhelper/retry_sender.go#L95).

Do you think we may miss anything here that would necessitate including information about the failed consumer in the error?

Reply:
The fanout doesn't have a state, it just responds with an error (or nil) upstream once all the downstream consumers have responded. Passing information (as part of the joined error) about what consumer failed is one problem, another problem is to use that information by a component that is going to do the retry and pass it back to the fanout (likely with context) so the fanout can retry only the failed destination.

There is nothing to do to address it in this PR. My idea is that we need to think about the potential expansion of the error type to add that information in the future.

Reply:
I added `consumererror.Component` to communicate this upstream in the pipeline.